src/Parallel/ResourceInfo.hpp

```diff
@@ -186,4 +205,698 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
+ *       Proc: 2
+ *       Exclusive: true
+ *     MySingleton2:
+ *       Proc: Auto
+ *       Exclusive: false
+ * \endcode
+ *
+ * where `MySingleton1` is the `pretty_type::name` of the singleton component
+ * and the options for each singleton are described in
+ * `Parallel::SingletonInfoHolder`.
+ *
+ * If you have both `Parallel::Tags::AvoidGlobalProc0` and
+ * `Parallel::Tags::SingletonInfo` tags in the initialization tags, you can
+ * combine the blocks like so:
+ *
+ * \code {.yaml}
+ * ResourceInfo:
+ *   AvoidGlobalProc0: true
+ *   Singletons:
+ *     MySingleton1:
+ *       Proc: 2
+ *       Exclusive: true
+ *     MySingleton2:
+ *       Proc: Auto
+ *       Exclusive: false
+ * \endcode
+ *
+ * Several consistency checks are done during option parsing to avoid user
+ * error. However, some checks can't be done during option parsing because the
+ * number of nodes/procs is needed to determine if there is an inconsistency.
+ * These checks are done during runtime, just before the map of singletons is
+ * created.
+ *
+ * To automatically place singletons, we use a custom algorithm that will
+ * distribute singletons evenly over the number of nodes, and evenly over the
+ * procs on a node. This will help keep communication costs down by distributing
+ * the workload over all of the communication cores (one comm core per charm
+ * node), and ensure that our resources are being maximally utilized (i.e. one
+ * core doesn't have all the singleton on it).
+ *
+ * Defining some terminology for singletons: `requested` means that a specific
+ * processor was requested in the input file; `auto` means that the processor
+ * should be chosen automatically by the algorithm; `exclusive` means that no
+ * other singletons or array elements should be placed on this singletons
+ * processor; `nonexclusive` means that you *can* place other singletons or
+ * array elements on this singletons processor. The algorithm is as follows:
+ *
+ * 1. Allocate all singletons that requested specific processors, both exclusive
+ *    and nonexclusive. This is done during option parsing.
+ * 2. Allocate auto exclusive singletons, distributing the total number of
+ *    exclusive singletons (auto + requested) as evenly as possibly over the
+ *    number of nodes. We say "as evenly as possible" because this depends on
+ *    the requested exclusive singletons. For example, if we have 4 nodes and 5
+ *    cores per node, the number of requested exclusive singletons on each node
+ *    is (0, 1, 4, 1), and we have 3 auto exclusive singletons to place, the
+ *    best distribution of exclusive singletons we can achieve given our
+ *    constraints is (2, 2, 4, 1). Clearly this is not the *most* evenly
+ *    distributed the exclusive singletons could be. However, this *is* the most
+ *    evenly distributed they could be given the starting distribution from the
+ *    input file.
+ *    3. Allocate auto nonexclusive singletons, distributing the total number of
```

Comment:
This isn't formatting correctly, not sure why.  (It is coming out as 2a instead of 3.)

Reply:
Strange...it looked correct for me locally. I fixed the indent, hopefully that helps.

Reply:
Ok yeah that looked like it fixed it.

Reply:
It may be version-dependent or something.  I occasionally find that my local version and the CI generated docs disagree.